syntax/decode.go

```diff
@@ -121,34 +122,52 @@ func unmarshalerDecoder(d *decodeState, v reflect.Value, opts decOpts) int {
 //////////
 
 func uintDecoder(d *decodeState, v reflect.Value, opts decOpts) int {
-	var uintLen int
-	switch v.Elem().Kind() {
-	case reflect.Uint8:
-		uintLen = 1
-	case reflect.Uint16:
-		uintLen = 2
-	case reflect.Uint32:
-		uintLen = 4
-	case reflect.Uint64:
-		uintLen = 8
+	if opts.varint {
+		return varintDecoder(d, v, opts)
 	}
 
-	buf := make([]byte, uintLen)
-	n, err := d.Read(buf)
-	if err != nil {
-		panic(err)
-	}
-	if n != uintLen {
+	uintLen := int(v.Elem().Type().Size())
+	buf := d.Next(uintLen)
+	if len(buf) != uintLen {
 		panic(fmt.Errorf("Insufficient data to read uint"))
 	}
 
+	return setUintFromBuffer(v, buf)
+}
+
+func varintDecoder(d *decodeState, v reflect.Value, opts decOpts) int {
+	// Read the first octet and decide the size of the presented varint
+	first := d.Next(1)
+	if len(first) != 1 {
+		panic(fmt.Errorf("Insufficient data to read varint length"))
+	}
```

Comment:
Just to confirm, this panic gets caught above, right?

Reply:
Yep.  This follows the general pattern (from `encoding/json`) where there's a top-level `recover()` and everything else uses `panic()`.